### ISSUE TITLE
Shadow tree encapsulation and CSS

### ISSUE THREAD

**Issue by annevk** [WICG/webcomponents#804]
I'd like to open this as a tracking issue to figure out to what extent CSS and Selectors violate the encapsulation principles we agreed on and what can be done about it. If you know of any such issues please add them here.

A clear one to start with is `:focus-within`. If `:focus` matches and `:focus-within` does not, there's a shadow tree. (To be clear, the mere fact of exposing that there is a shadow tree violates encapsulation, as previously established.)

cc @emilio @tabatkins @fantasai

**Comment by emilio** [comment]
Well, to be clear, the condition here is that `:focus-within` matches, `:focus` does not, and `document.activeElement` is the element you're testing against.

**Comment by tabatkins** [comment]
There is no case in which :focus matches but :focus-within doesn't; that's impossible, as :focus-within is defined as a superset of :focus. Did you mean the reverse, where an element matches :focus-within but none of the inclusive descendants match :focus?

**Comment by emilio** [comment]
Yeah, that's it.

**Comment by tabatkins** [comment]
What is returned by `document.activeElement` when the focus is in a shadow tree? Does open/closed matter?

**Comment by emilio** [comment]
It's retargeted, it doesn't matter whether it's an open or closed ShadowRoot.

**Comment by tabatkins** [comment]
If it's retargeted, doesn't that affect `:focus`, too? The Selectors definition of `:focus` is deliberately a little vague to allow host languages some wiggle room in defining it.

**Comment by emilio** [comment]
Right now it does not in any browser. It could be defined to be like that and make `:focus` match for all containing shadow hosts, but that wouldn't be a reasonable definition, you need to account for the cascade origin as well at least. See:

  https://groups.google.com/a/chromium.org/d/msg/blink-dev/X-qPSmdSR_g/kZ-4UHdpAwAJ

/cc @hayatoito

**Comment by tabatkins** [comment]
Are you referring to the fact that UA sheets match in all shadow contexts? (Thus a retargeted :focus rule would match on every shadow host up to the outer light DOM, when an element in a deeply-nested component was focused.)

That could be taken care of, I suppose.

**Comment by tabatkins** [comment]
But note then that the issue in Anne's original comment is thus further wrong; there's nothing wrong with :focus-within, as it exposes no information not already exposed via JS (the retargeted `.activeElement`), but rather this is an *old* issue with focus (more generally, not just the `:focus` pseudo-class) and shadow DOM, which has been known for years.

(In other words, it's the pre-existing fact that `:focus` and `.activeElement` can mismatch that reveals there's a shadow tree. `:focus-within` doesn't bring anything new here.)

**Comment by annevk** [comment]
That is a good point, I should not have filed this in haste. If both `:focus` and `:focus-within` were to match on shadow hosts, as they do on other elements that have no (node tree) children and are focusable (somehow), it'd be fine. It's unclear if that's doable for `:focus` at this point of course. (Apart from the default style sheet problem @emilio raised.)

The thing that's true for `:focus` is also a question for `:fullscreen` and I suppose this new `:picture-in-picture` pseudo-class.

The other issue I was thinking of when filing this is https://github.com/whatwg/html/issues/3748.

**Comment by hayatoito** [comment]
FYI.

- `:focus` is defined at https://w3c.github.io/webcomponents/spec/shadow/#focus
- It looks there is an filed issue: https://github.com/whatwg/html/issues/2013
- @rakina took over focus related works from the former google who was working on this.

**Comment by rakina** [comment]
> @rakina took over focus related works from the former google who was working on this.

FYI, I'm not working on this area anymore. cc @chrishtr who's taking over things related to this.

**Comment by Jamesernator** [comment]
Most of the new Houdini APIs violate encapsulation by introducing new global namespaces for their registered things:

- `CSS.animationWorklet`
- `CSS.paintWorklet`
- `CSS.layoutWorklet`
- `CSS.registerProperty`

There's also [this issue](https://github.com/w3c/csswg-drafts/issues/1995) with has additional ones like `@font-face`/`@keyframes`.

**Comment by rniwa** [comment]
FWIW, having any observable behavioral difference on an element when it has a shadow root is bad so we should probably try to make `:focus` match a shadow host if the shadow root contains a focused element. In general, everything related to focus & Shadow DOM needs to be studied & clarified. There is a lot of missing pieces there.

**Comment by annevk** [comment]
@Jamesernator that doesn't necessarily violate encapsulation though. Can they be used to reveal that something has a shadow tree?

**Comment by Jamesernator** [comment]
Hmmm I thought worklets had a `postMessage` in them but I must be confusing them with other types of workers. Given this I don't think a worklet can communicate a message out to tell the main thread that it's methods were invoked (other than timing-based messages) which could reveal some shadow trees.

**Comment by Jamesernator** [comment]
What constitutes "CSS and Selectors violate the encapsulation principles"? 

Because it seems to me that the `:focus`/`:focus-within` doesn't really violate encapsulation as in order for `:focus`/`:focus-within` to detect a shadow root you actually need to add a `:focus`-able element inside the tree within the Shadow root for this to occur.

Similarly can't we detect *all* shadow roots with no default slot simply by appending a child with size to an element we want to check? If the CSSOM is different than what we expected (e.g. `clientWidth`, `getComputedStyle`, etc) then there's a shadow root.

Example of detecting shadow roots without a default slot:

```html
<!doctype html>

<div id="hasShadowRoot"></div>
<div id="hasNoShadowRoot"></div>

<script>
  const shadowRoot = hasShadowRoot.attachShadow({ mode: 'closed' });

  function doesHaveShadowRoot(el) {
    if (!el.isConnected) {
      document.body.appendChild(el)
    }
    const child = document.createElement('div');
    child.style.width = '10px';
    child.style.height = '10px';

    document.body.appendChild(child);
    const {
      clientHeight: predictedHeight,
      clientWidth: predictedWidth,
    } = child;

    el.appendChild(child);
    const {
      clientHeight: realHeight,
      clientWidth: realWidth,
    } = child;

    child.remove():

    if (predictedHeight !== realHeight || predictedWidth !== realWidth) {
      return 'definitely has shadow root'
    }
    return 'not sure if has shadow root'
  }

  document.body.insertAdjacentHTML('beforeend', `
    hasShadowRoot: ${ doesHaveShadowRoot(hasShadowRoot) }
    <br>
    hasNoShadowRoot: ${ doesHaveShadowRoot(hasNoShadowRoot) }
  `)
</script>
```

It seems to me that the only shadow roots that can be truly encapsulated from detection are those that don't affect layout of the child elements in any way (having a child that is `:focus`-able changes layout so would be detectable anyway).

**Comment by chrishtr** [comment]
I don't think that :picture-in-picture, :fullscreen or :focus should automatically apply to the shadow host chain. Would be good to discuss this at the F2F next week also.

Thoughts about why:

* :focus doesn't do that (instead there is a delegatesFocus shadow root attribute to opt into this behavior)
* it doesn't make sense for :fullscreen either because of the injected stylesheet issue Emilio mentioned
* activeElement already can be present even if no element in the current document has :focus, so having that true for pictureInPictureElement or fullscreenElement is no "worse"
* Retargeting violates the principle that shadow roots should try to have observable-implementation and performance encapsulation

For :picture-in-picture specifically, I can think of:

* It could be unexpected (and is not controllable) for some developers to have retargeting result in styling of multiple contained shadow hosts that have different stylesheet contexts

**Comment by rniwa** [comment]
> * :focus doesn't do that (instead there is a delegatesFocus shadow root attribute to opt into this behavior)

That's simply not true. If an element inside a shadow tree has a focus, then the shadow host matches `:focus`. We specifically added `:focus-visible` which doesn't apply to shadow host chain. See https://jsfiddle.net/z9jkfcLq/1/ for an example.
